### PR TITLE
Add custom x axes to TensorBoard

### PR DESCRIPTION
**Pull request by naeioi**
This PR adds support for custom x-axes to the TensorBoard scalar plot. The axis names are specified in the tensorboard output constructor. 
```
class TensorBoardOutput(LogOutput):
    def __init__(self,
                 log_dir,
                 x_axes=None,
                 flush_secs=120,
                 histogram_samples=1e3): 
```

When `x_axes` is `None`, it falls back to use iterations as the x-axis. 
If any x_axis is not present in the scalar tabular. A warning will be logged to the console.
If all x_axes are not present, it falls back to use iteration as the x-axis.

Screenshots of an experiment with `Epoch` and `TotalEnvSteps` as x-axes.
Screenshot from 2019-11-22 20-43-24
Screenshot from 2019-11-22 20-43-43

I will open a separate PR in garage repo to set `TotalEnvSteps` as the default x-axis.

### CODE REVIEW

**Comment by ryanjulian** [comment]
> If any x_axis is not present in the scalar tabular. A warning will be logged to the console.

I think this should be an error -- you told me to expect a key and then didn't give it to me.

**Comment by ryanjulian** [comment]
I'm actually willing to endure the API breakage of not implicitly counting itrs in `TensorBoardOutput` -- because I think that it's a bad pattern.

@krzentner WDYT?

File: tests/dowel/test_tensor_board_output.py [hunk 1]
```diff
@@ -13,18 +13,25 @@
...
+        self.tensor_board_output = TensorBoardOutput(
+            self.log_dir.name, x_axes=self.x_axes)
 
     def tearDown(self):
         self.tensor_board_output.close()
         self.log_dir.cleanup()
         super().tearDown()
 
+    @classmethod
+    def setXAxes(cls, x_axes):
```
Comment (ryanjulian) [inline tests/dowel/test_tensor_board_output.py]: please use `snake_case`

File: src/dowel/logger.py [hunk 2]
```diff
@@ -267,15 +267,21 @@ def dump_output_type(self, output_type, step=None):
...
     def dump_all(self, step=None):
         """Dump all outputs connected to the logger.
 
         :param step: The current run step.
         """
         for output in self._outputs:
-            output.dump(step=step)
+            try:
+                output.dump(step=step)
+            except LoggerWarning as e:
```
Comment (ryanjulian) [inline src/dowel/logger.py]: why not let the output just print the warning itself? Warnings are not exceptions -- they don't break flow control.

File: src/dowel/tensor_board_output.py [hunk 3]
```diff
@@ -106,3 +132,20 @@ def dump(self, step=None):
     def close(self):
         """Flush all the events to disk and close the file."""
         self._writer.close()
+
+
+class NonexistentAxesError(LoggerWarning):
```
Comment (ryanjulian) [inline src/dowel/tensor_board_output.py]: if these are going to be warnings, they should not be raised as exceptions. You can use `warnings.warn` to print these without interrupting flow control.

File: src/dowel/tensor_board_output.py [hunk 4]
```diff
@@ -66,10 +72,30 @@ def record(self, data, prefix=''):
...
+        custom_axes = custom_axes and not (len(nonexist_axes) == len(
+            self._x_axes))
+
         for key, value in data.as_dict.items():
-            self._record_kv(key, value, step)
+            if isinstance(value, np.ScalarType) and custom_axes:
+                for axis in self._x_axes:
+                    if axis not in nonexist_axes and key is not axis:
+                        x = data.as_dict[axis]
+                        self._record_kv('{}/{}'.format(axis, key), value, x)
```
Comment (ryanjulian) [inline src/dowel/tensor_board_output.py]: if you were going to do this, you should use the `prefix` feature.

I recommend the following pattern, though, which I think will lead to better TensorBoard displays:
* Default x-axis (either specific or just `x_axes[0]`) gets unmodified keys, such as `Policy/Entropy`
* Additional x-axes get keys pre- or post-fixed with some non-`/` char and their x-axis

e.g.:
* post: : `Policy/Entropy@Itr`
* pre: `Itr@Policy/Entropy`

I recommend you try both and see which is most intuitive for someone who doesn't know how to write a regex.

These are intended to:
* Make it easy to filter the TB display to show only one kind of x-axis using regexes
* Preserve the sections in the web-app, which are defined by the first level of `/` tokens

**Comment by ryanjulian** [comment]
If you want to share a sample with us, you can use https://tensorboard.dev

File: src/dowel/logger.py (hunk 2 above)
Comment (naeioi) [reply src/dowel/logger.py]: Three reasons,
- An output will need a `self._warn()` to print the warning itself, either by making its own, or borrowing from the `Logger`. If it makes its own, the `Logger` loses control of warning printing (and there will be duplicate code). If it borrows from the `Logger`, it will require the `Logger` instance passed to the `Output`. 
- By raising a warning exception, an output asks the logger to handle the printing at its discretion. 
- Warnings are subclasses of exception. https://docs.python.org/3/library/warnings.html#warning-categories. And I see a warning class defined in the logger. https://github.com/rlworkgroup/dowel/blob/621d3e421f9553dde474d6d6ba82c91ff2da6821/src/dowel/logger.py#L328 I think this suggests material "sub"-warnings in outputs.

File: src/dowel/tensor_board_output.py (hunk 4 above)
Comment (naeioi) [reply src/dowel/tensor_board_output.py]: >  Default x-axis (either specific or just `x_axes[0]`) gets unmodified keys, such as `Policy/Entropy`

This is probably not a good idea, because the axis name is not displayed on tensorboard.

File: src/dowel/logger.py (hunk 2 above)
Comment (ryanjulian) [reply src/dowel/logger.py]: These are excellent points, but don't change the fact using exceptions for flow control is a big problem. Furthermore, it's simply not idiomatic python to use `raise` with warnings. That's why the `warnings` module exists.

When they do happen, it's likely that this warnings will be raised repeatedly. Exceptions break most optimizations in interpreters and modern CPUs. Firing a bunch of them repeatedly for flow control can silently hose performance.

`self_warn()` exists so we can only print warnings once. I think you have run afoul of this TODO: https://github.com/rlworkgroup/dowel/issues/20.

I suggest you either make a process-global `warn_once` or just spam the user with warnings.

File: src/dowel/tensor_board_output.py (hunk 4 above)
Comment (naeioi) [reply src/dowel/tensor_board_output.py]: > * post: : `Policy/Entropy@Itr`
> * pre: `Itr@Policy/Entropy`

I like the post one. The pre one will double the number of sections if there are two axes, which is visually distracting.

Comment (ryanjulian) [reply src/dowel/tensor_board_output.py]: *re: defaults*
Most users will be using one x-axis key, and the default should not be ugly. Perhaps you can accept both a `str` and `list` for `x_axes` as courtesy to users and only apply labels in the `list` case. I still think this isn't quite right, because even users of more than one x-axis have a default in mind which they will look at 90% of the time, and only use the others for debugging specific issues.

Alternatively, you could change the API  to have 2 args: `x_axis` and `additional_x_axes`, and only label the for `additional_x_axes`. This is a little less magical than what I suggested with using `x_axes[0]`.

As of today, we also don't have axis labels and the default is implicitly `Itr` but noted nowhere, so we already have the problem you mentioned today. AFAIK nobody's ever asked me what our x-axis represents. If you are scared of not exposing necessary information, you could always output a `Text` summary which tells the user how TensorBoardOutput is configured. In practice, I doubt that. This is a basic flaw in the TensorBoard interface, so we're not going to be able to solve it fully.

*re:pre- vs post-fix*
It's all about which distraction you prefer. It might be better to have all of the plots of one x-axis type grouped into their own section, so you can look at just those without using a regex.

Perhaps you can make examples of both for us to play with?

File: src/dowel/logger.py (hunk 2 above)
Comment (naeioi) [reply src/dowel/logger.py]: I don't feel like fixing #20 here - that will extend the scope of this PR. I will follow `CsvOutput`'s pattern to implement a `self._warn()` in `TensorboardOutput`.

File: src/dowel/tensor_board_output.py (hunk 4 above)
Comment (naeioi) [reply src/dowel/tensor_board_output.py]: _re: defaults_
When the axis name is not displayed, printing this name to console helps a bit, but will easily be omitted. There are simply too many messages on the console and people will only scrutinize them when the program fails.

TensorBoard has the flaw of not displaying the x-axis name, but your suggestion of using `@axis` is an excellent workaround to hint people what they are reading. 

Plus, the default x-axis is specified in `experiment_wrapper.py`, which is a file that most regular users will not touch. 

Printing this name to TensorBoard will add some redundancy, but I think it wins by being much easier to understand.

Comment (ryanjulian) [reply src/dowel/tensor_board_output.py]: `@axis` is nice for non-defaults but it makes the default case worse, which is the case which 90% of users will only ever experience. this is pretty unpythonic. see: https://www.python.org/dev/peps/pep-0008/#a-foolish-consistency-is-the-hobgoblin-of-little-minds

this intention here is to establish that the default x-axis in garage is always `TotalEnvSteps`. we can put it in the documentation and all over the code. beating users over the head with it over and over again in the TensorBoard is consistent in the way that makes programmers feel "correct" and all warm and fuzzy inside, and makes life harder for everyone else for no reason.

i'll add more reviewers to ask what they think. if you really want to convince people your way is better then make some demos. you are proposing changing the status quo, so the burden is on you to convince everyone.

Comment (naeioi) [reply src/dowel/tensor_board_output.py]: I will take your suggestion of using `x_axis` + `additional_x_axes` as `TensorboardOutput`'s interface.

Meanwhile, I want to point out that `TotalEnvSteps` as x-axis is not status quo in open-source frameworks' support for Tensorboard. Most frameworks use iterations, just like what garage is doing now. To list some of the most popular ones,

[Dopamine](https://github.com/google/dopamine/blob/6463bfa8660daf17823825ab884b118d3a57ea4e/dopamine/agents/dqn/dqn_agent.py#L425)
[tf-agents](https://github.com/tensorflow/agents/blob/866500aea940f27f6660bd16eed7acb339377d37/tf_agents/agents/ppo/examples/v2/train_eval.py#L261)
ray-rllib, [code1](https://github.com/ray-project/ray/blob/7f8de6144172fbc60fc91b67f5dcaad183424058/python/ray/tune/logger.py#L189), [code2](https://github.com/ray-project/ray/blob/7f8de6144172fbc60fc91b67f5dcaad183424058/python/ray/tune/logger.py#L214)
[keras-rl](https://github.com/keras-rl/keras-rl/blob/216c3145f3dc4d17877be26ca2185ce7db462bad/rl/core.py#L234)

So after this PR, garage will be one of the few frameworks, if not the only one, that support using `TotalEnvSteps` as the x-axis in TensorBoard.

Comment (ryanjulian) [reply src/dowel/tensor_board_output.py]: Yes, and IMO it will be a great differentiator. One of the great hassles of using RL libraries is that the x-axis of the output logs differ greatly between libraries when the mathematically-valid x-axis is always TotalEnvSteps (or perhaps TotalGradSteps).

Publications use TotalEnvSteps almost universally to compare algorithms. Determining the actual number of TotalEnvSteps, if it's not explicitly logged, can itself be a challenge which takes hours to figure out in an unfamiliar codebase, depending on how obfuscated the code is.

btw, tf_agents uses the global gradient step counter (e.g. `TotalGradSteps`), not `Itr` as defined by garage. This can make a lot sense sometimes in off-policy algorithms, but is mostly a hold-over from the practices used for supervised learning. The relevant publication benchmark is still `TotalEnvSteps`.

File: src/dowel/tensor_board_output.py [hunk 5]
```diff
@@ -23,31 +24,49 @@
...
     def __init__(self,
                  log_dir,
-                 x_axes=None,
+                 x_axis=None,
+                 additional_x_axes=None,
                  flush_secs=120,
                  histogram_samples=1e3):
+        if x_axis is None:
+            assert not additional_x_axes, (
+                'You have to specify a x_axis if you want additional axes.')
```
Comment (ryanjulian) [inline src/dowel/tensor_board_output.py]: an

File: src/dowel/logger.py [hunk 6]
```diff
@@ -328,4 +328,6 @@ def disable_warnings(self):
 class LoggerWarning(UserWarning):
     """Warning class for the Logger."""
 
-    pass
+    @abc.abstractmethod
+    def to_string(self):
+        raise NotImplementedError
```
Comment (ryanjulian) [inline src/dowel/logger.py]: abstractmethods should just `pass` or leave the body empty

**Comment by naeioi** [comment]
I found that allowing the missing axis is inevitable. For example, DDPG calls `logger.dump()` every epoch, but the scalar table is empty for the beginning several epochs. So I keep this as a warning but not error.

I also found that '@' in the name will be replaced with underline by tensorboard. The followings are candidate naming formats. 

post_slash: https://tensorboard.dev/experiment/sydhawaaQr2BFlCnqzF2qQ/#scalars
post_underline: https://tensorboard.dev/experiment/wQsEzHaGQkekfnB5GqSz9g/#scalars
pre_slash: https://tensorboard.dev/experiment/zZi3idwRRGOq4SVOybs2Nw/#scalars
pre_underline: https://tensorboard.dev/experiment/Bt1o8m20S6Op5nXInfQgvA/#scalars
pre_full_axes_name: https://tensorboard.dev/experiment/jUElLlA3TQu7d7AskiBuvw/#scalars

**Comment by ryanjulian** [comment]
Ah, let me take a look at those. What a bummer that TF removes the @
